src/main/clojure/clara/rules/memory.cljc

```diff
@@ -241,7 +256,78 @@
+                     ;; pre-1.9.0.
+                     (if-let [^clojure.lang.ISeq entries (.seq ^clojure.lang.Seqable bindings)]
+                       ;; Type hint to Indexed vs MapEntry just because MapEntry seems to be a
+                       ;; less stable impl detail to rely on.
+                       (loop [^clojure.lang.Indexed entry (.first entries)
+                              entries (.next entries)]
+                         (let [k (some-> entry (.nth 0))
+                               v (some-> entry (.nth 1))]
+                           (if (and k
+                                    ;; other-bindings will always be persistent map so invoke
+                                    ;; it directly.  It is faster than `get`.
+                                    (compare-fn v (other-bindings k)))
+                             (recur (some-> entries .first)
+                                    (some-> entries .next))
+                             ;; If there is no k left, then every entry matched.  If there is a k,
+                             ;; that means the comparison failed, so the maps aren't equal.
+                             (not k))))
+                       ;; Empty bindings on both sides.
+                       true))) 
+
+              ;; Check the :matches on each token.  :matches need to be in the same order on both
+              ;; tokens to be considered the same.
+              (let [^clojure.lang.Indexed matches (:matches token)
+                    ^clojure.lang.Indexed other-matches (:matches other-token)
+                    count-matches (.size ^java.util.List matches)]
+                (and (= count-matches
+                        (.size ^java.util.List other-matches))
+                     (loop [i 0]
+                       (cond
+                         (= i count-matches)
+                         true
+
+                         ;; Compare node-id's first.  Fallback to comparing facts.  This will
+                         ;; most very likely be the most expensive part to execute.
+                         (let [^clojure.lang.Indexed m (.nth matches i)
+                               ^clojure.lang.Indexed om (.nth other-matches i)]
+                           ;; A token :matches tuple is of the form [fact node-id].
+                           (and (= (.nth m 1) (.nth om 1))
+                                (compare-fn (.nth m 0) (.nth om 0))))
+                         (recur (inc i))
+
+                         :else
+                         false))))))))
+
+#?(:clj
+   (defprotocol IdentityComparable
+     (using-token-identity [this bool])))
+
+#?(:clj
+   (deftype RuleOrderedActivation [node-id
+                                   token
+                                   activation
+                                   rule-load-order
+                                   ^:unsynchronized-mutable use-token-identity?]
```

Comment:
I added a comment above `using-token-identity!` describing this situation.  It isn't ideal that this even needs a comment, but I think it is fine and works for now.

I also changed the name `using-token-identity` to `using-token-identity!` since the function most certainly would have to update some sort of mutable state to be useful.  I missed this convention on the first-pass.

Let me know if you have any further issues with this right now.